wheel/python/clvm_rs/program.py

```diff
@@ -0,0 +1,346 @@
+        `
+        if p.at_many("f", "rf", "rfrf") == [5, 10, 15]:
+        `
+        instead of
+        `
+        if [p.at("f"), p.at("rf"), p.at("rfrf")] == [5, 10, 15]:
+        `
+        """
+        return [self.at(_) for _ in positions]
+
+    def replace(self, **kwargs) -> "Program":
+        """
+        Create a new program replacing the given paths (using `at` syntax).
+        Example:
+        ```
+        >>> p1 = Program.to([100, 200, 300])
+        >>> print(p1.replace(f=105) == Program.to([105, 200, 300]))
+        True
+        >>> p2 = [100, 200, [301, 302]]
+        >>> print(p1.replace(rrf=[301, 302]) == Program.to(p2))
+        True
+        >>> p2 = [105, 200, [301, 302]]
+        >>> print(p1.replace(f=105, rrf=[301, 302]) == Program.to(p2))
+        True
+        ```
+
+        This is a convenience method intended for use in the wallet or
+        command-line hacks where it would be easier to morph elements
+        of an existing clvm object tree than to rebuild one from scratch.
+
+        Note that `Program` objects are immutable. This function returns a
+        new object; the original is left as-is.
+        """
+        return self.to(replace(self, **kwargs))
+
+    def tree_hash(self) -> bytes:
+        # we operate on the unwrapped version to prevent the re-wrapping that
+        # happens on each invocation of `Program.pair` whenever possible
+        if self._cached_sha256_treehash is None:
+            self._cached_sha256_treehash = sha256_treehash(self._unwrapped)
+        return self._cached_sha256_treehash
+
+    def run_with_cost(
+        self, args, max_cost: int = MAX_COST, flags: int = 0
+    ) -> Tuple[int, "Program"]:
+        prog_bytes = bytes(self)
+        args_bytes = bytes(self.to(args))
+        try:
+            cost, lazy_node = run_serialized_chia_program(
+                prog_bytes, args_bytes, max_cost, flags
+            )
+            r = self.wrap(lazy_node)
+        except ValueError as ve:
+            raise EvalError(ve.args[0], self.wrap(ve.args[1]))
+        return cost, r
+
+    def run(self, args) -> "Program":
+        cost, r = self.run_with_cost(args, MAX_COST)
+        return r
+
+    def curry(self, *args) -> "Program":
+        """
+        Replicates the curry function from clvm_tools, taking advantage of *args
+        being a list.  We iterate through args in reverse building the code to
+        create a clvm list.
+
+        Given arguments to a function addressable by the '1' reference in clvm
+
+        fixed_args = 1
+
+        Each arg is prepended as fixed_args = (c (q . arg) fixed_args)
+
+        The resulting argument list is interpreted with apply (2)
+
+        (a (q . self) rest)
+
+        Resulting in a function which places its own arguments after those
+        curried in in the form of a proper list.
+        """
+        return self.to(self.curry_treehasher.curry(self, *args))
+
+    def uncurry(self) -> Tuple[Program, Optional[List[Program]]]:
```

Comment:
I don't have a good understanding of how repeated attempts at uncurrying something would be done. Would it be to determine what type of contract a puzzle is?

Seems to me that there should be multiple layers, one which uncurries the puzzle to determine the `mod`, then another that dispatches to the appropriate sub-wallet handler based on that mod, passing in the curried parameters at that time to obviate the need for a second uncurrying.

Reply:
> Seems to me that there should be multiple layers, one which uncurries the puzzle to determine the mod, then another that dispatches to the appropriate sub-wallet handler based on that mod, passing in the curried parameters at that time to obviate the need for a second uncurrying

Right yeah, this is what we had to build to solve the problem so I'm wondering if we could eliminate the need for this architecture (and eliminate the possibility of omitting it and writing inefficient code because you don't know that `uncurry` is inefficient) by caching the values

Reply:
How about this code flow: if you know a puzzle is a curried `MOD`, uncurry it to get the `MOD` and parameters. You use the `MOD` hash to figure out which "driver", then the curried parameters to figure out which instance of the driver is appropriate.

I think a list of `if MOD == a`, `if MOD == b`, etc. is slow. In general, a `switch/case` can often be replaced with a `dict` look-up.